[#23] 회원정보의 수정에 대한 기능 구현

donsonioc2010 opened the pull request:
## 기능 정리 
1. `UserInfoUpdateDto`에서 수정가능한 권한으로 Validation진행
2. 수정이 가능한 경우 DB Update실행 이후, 수정 결과에 대하여 결과값을 Return

donsonioc2010 committed:
refactor : 패스워드 변경기능 수정

- 파라미터 Null인 경우 null로 다시 return

donsonioc2010 committed:
feat : 사용자 정보 수정기능 구현

donsonioc2010 committed:
refactor : 조건 수정

donsonioc2010 committed:
test : 유저 정보 수정 기능에 대한 Test

donsonioc2010 committed:
refactor : validation수정

donsonioc2010 committed:
Merge Develop

donsonioc2010 committed:
Merge와 Renmae으로 발생한 오탈자 수정

donsonioc2010 committed:
refactor : 수정내역 상세 메세지 참고

- 정규식 Constants참고하도록 수정
- 주석 인터페이스로 위치 변경
- SQLException를 통한 분기처리 제거

f-lab-kyle commented on src/main/java/com/nooblol/community/controller/UserController.java:
```diff
@@ -27,11 +29,26 @@ public class UserController {
+   * @return 정상적으로 Update가 성공했으면 OK Response와 결과값으로 true가 Return되며, 정보 수정에 실패하면 OK상태코드와 false값을
+   * Return한다
+   */
+  @PostMapping("/update")
```
POST /user 자체가 update를 의미하지 않을까요?

f-lab-kyle commented on src/main/java/com/nooblol/community/service/impl/UserInfoServiceImpl.java:
```diff
@@ -0,0 +1,57 @@
+      }
+      return result;
+    } catch (Exception e) {
+      throw new IllegalArgumentException(ExceptionMessage.SERVER_ERROR);
```
에러가 발생하면 원인이 무엇인지 알 수 없을것 같아요  (e 로깅 해야한단 이야기)

f-lab-kyle commented on src/main/java/com/nooblol/community/service/impl/UserInfoServiceImpl.java:
```diff
@@ -0,0 +1,57 @@
+      );
+
+      ResponseDto result = ResponseEnum.OK.getResponse();
+      if (userInfoMapper.updateUserInfo(userUpdateInfoDto) == 0) {
+        result.setResult(false);
+      } else {
+        result.setResult(true);
+      }
```
```suggestion
        result.setResult(userInfoMapper.updateUserInfo(userUpdateInfoDto) != 0);
```

f-lab-kyle commented on src/main/java/com/nooblol/community/service/impl/UserInfoServiceImpl.java:
```diff
@@ -0,0 +1,57 @@
+          EncryptUtils.stringChangeToSha512(userUpdateInfoDto.getOrgPassword())
+      );
+
+      //New Password의 경우 null이 존재 할 수 있다. 쿼리문에서 분기처리가 존재하기에 null이 와도 상관 없다.
```
null이면 구지 쿼리까지 안가고 처리해도 되지 않을까요?
더불어, 쿼리 레벨에서 null체크를 하는건 책임이 분산되는 느낌도 들고요

f-lab-kyle commented on src/main/java/com/nooblol/global/utils/EncryptUtils.java:
```diff
@@ -20,6 +21,9 @@ public class EncryptUtils {
    */
   public static String stringChangeToSha512(String value)
```
null을 리턴하는 메소드들은 가급적 Optional 을 활용해보세요. 메소드를 사용하는 '코드레벨' 에서 null 을 체크 해야 하는구나 라고 '인지' 하는 효과가 생기거든요.

f-lab-kyle commented on src/main/resources/mybatis/mapper/community/UserInfoMapper.xml:
```diff
@@ -0,0 +1,20 @@
+  <update id="updateUserInfo" parameterType="UserInfoUpdateDto">
+    UPDATE users
+    SET updated_at = #{updatedAt}
+    <if test="newUserName != null and newUserName != ''">
+      , user_name = #{newUserName}
+    </if>
+    <if test="newPassword != null and newPassword != ''">
+      , user_password_hash = #{newPassword}
+    </if>
```
위에서 이야기 했는데, 여기서 체크를 하면(둘다 비어있으면) 실제로 업데이트를 하지 않았는데 업데이트 일자만 변경되니 문제(?)가 되지 않을까요?

donsonioc2010 committed:
refactor : Review적용

- Mapper에서 Null체크를 하는 대신에 값이 없는 경우에는 이전의 값을 그대로 가져가도록 수정
- catch문으로 감싸다보니 e로깅이 불가능한 문제 수정
- EncryptUtils Optional활용
